src/mlpack/methods/linear_svm/linear_svm_impl.hpp

```diff
@@ -0,0 +1,163 @@
+/**
+ * @file linear_svm.cpp
+ * @author Ayush Chamoli
+ *
+ * Implementation of Linear SVM.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_IMPL_HPP
+#define MLPACK_METHODS_LINEAR_SVM_LINEAR_SVM_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "linear_svm.hpp"
+
+namespace mlpack {
+namespace svm {
+
+template <typename MatType>
+template <typename OptimizerType>
+LinearSVM<MatType>::LinearSVM(
+    const MatType& data,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    const double lambda,
+    OptimizerType optimizer) :
+    numClasses(numClasses),
+    lambda(lambda)
+{
+  Train(data, labels, numClasses, lambda, optimizer);
+}
+
+template <typename MatType>
+template <typename OptimizerType>
+double LinearSVM<MatType>::Train(
+    const MatType& data,
+    const arma::Row<size_t>& labels,
+    const size_t numClasses,
+    const double lambda,
+    OptimizerType optimizer)
+{
+  LinearSVMFunction<MatType> svm(data, labels,
+      numClasses, lambda);
+  if (parameters.is_empty())
+    parameters = svm.InitialPoint();
+
+  // Train the model.
+  Timer::Start("linear_svm_optimization");
+  const double out = optimizer.Optimize(svm, parameters);
+  Timer::Stop("linear_svm_optimization");
+
+  Log::Info << "LinearSVM::LinearSVM(): final objective of "
+            << "trained model is " << out << "." << std::endl;
+
+  return out;
+}
+
+template <typename MatType>
+void LinearSVM<MatType>::Classify(
+    const MatType& data,
+    arma::Row<size_t> &labels)
+const
```

Comment:
Could you call the other implementation of `Classify()` here instead of reimplementing?

Reply:
I followed these signature and the way they were implemented from Softmax Regression.
Should a similar thing be done for `LinearSVMFunction::Gradient()`, `LinearSVMFunction::Evaluate()` and `LinearSVMFunction::EvaluateWithGradient()` (Separable and non-separable) since the implementation is almost similar (I think not).

Reply:
There may be an efficiency improvement for `EvaluateWithGradient()` if any of the computation is shared.  You can take a look at the implementation to see.  If not, there's no need for `EvaluateWithGradient()`.

More details here: http://ensmallen.org/docs.html#differentiable-functions